### PR TITLE
shorter handle representation.

### PR DESCRIPTION
**Pull request by willscott**
one of two parts of #66

### CODE REVIEW

**Comment by ryscheng** [comment]
Is this ready for review or are we waiting for a part 2 of 2?

**Comment by willscott** [comment]
the trust domains in client.conf should probably also get hex treatment rather than the go json encoder default treatment of array of bytes.

That's on my list to do with #7 - having the connection between the client and frontend be over TLS with a pinned server cert.

**Comment by willscott** [comment]
which is to say. this one is ready for review.

File: libtalek/handle.go [hunk 1]
```diff
@@ -184,3 +185,41 @@ func (h *Handle) retrieveResponse(args *common.ReadArgs, reply *common.ReadReply
...
+		return nil, err
+	}
+	txt := fmt.Sprintf("%x.%x.%x.%x.%d", s1, s2, *h.SharedSecret, *h.SigningPublicKey, h.Seqno)
+	return []byte(txt), nil
+}
+
+// UnmarshalText restores a handle from its compact textual representation
+func (h *Handle) UnmarshalText(text []byte) error {
+	var s1, s2, ss, pk []byte
+	if n, err := fmt.Sscanf(string(text), "%x.%x.%x.%x.%d", &s1, &s2, &ss, &pk, &h.Seqno); n < 4 || err != nil {
```
Comment (ryscheng) [inline libtalek/handle.go]: n<5?

File: libtalek/handle_test.go [hunk 2]
```diff
@@ -39,6 +40,26 @@ func TestGeneratePoll(t *testing.T) {
...
+		t.Fatalf("Error serializing: %v\n", err)
+	}
+	fmt.Printf("Serialized handle looks like %s\n", txt)
+
+	h2, _ := NewHandle()
+	err = h2.UnmarshalText(txt)
+	if err != nil {
+		t.Fatalf("Could not deserialize: %v\n", err)
+	}
+	if !bytes.Equal(h.SharedSecret[:], h2.SharedSecret[:]) {
```
Comment (ryscheng) [inline libtalek/handle_test.go]: Should check other fields?

File: libtalek/topic.go [hunk 3]
```diff
@@ -111,3 +113,21 @@ func (t *Topic) encrypt(plaintext []byte, nonce *[24]byte) ([]byte, error) {
 	digest := ed25519.Sign(t.SigningPrivateKey, buf)
 	return append(buf, digest[:]...), nil
 }
+
+// MarshalText is a compact textual representation of a topic
```
Comment (ryscheng) [inline libtalek/topic.go]: Test?

**Comment by willscott** [comment]
addressed comments

File: libtalek/handle_test.go [hunk 4]
```diff
@@ -55,8 +55,8 @@ func TestSerialization(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Could not deserialize: %v\n", err)
 	}
-	if !bytes.Equal(h.SharedSecret[:], h2.SharedSecret[:]) {
-		t.Fatalf("serialization log info!")
+	if !bytes.Equal(h.SharedSecret[:], h2.SharedSecret[:]) || !bytes.Equal(h.SigningPublicKey[:], h2.SigningPublicKey[:]) {
```
Comment (ryscheng) [inline libtalek/handle_test.go]: What about the Seeds? Also I noticed that handle.SeqNo is not included in the Marshalling. Is this an oversight?

Comment (willscott) [inline libtalek/handle_test.go]: seeds have tests. seqNo is 0, in these tests so doesn't matter. The assumption is just that sscanf fills, i'm not convinced we gain much more confidence with the extra verbosity

File: libtalek/topic_test.go [hunk 5]
```diff
@@ -60,6 +62,22 @@ func TestSerializeRestore(t *testing.T) {
...
+		t.Fatalf("Error serializing: %v\n", err)
+	}
+	fmt.Printf("Serialized topic looks like %s\n", txt)
+
+	clone = Topic{}
+	err = clone.UnmarshalText(txt)
+	if err != nil {
+		t.Fatalf("Could not deserialize: %v\n", err)
+	}
+	if !bytes.Equal(topic.SigningPrivateKey[:], clone.SigningPrivateKey[:]) || !bytes.Equal(topic.Handle.SharedSecret[:], clone.Handle.SharedSecret[:]) {
```
Comment (ryscheng) [inline libtalek/topic_test.go]: Again, we're missing checks on some fields here. It seems like the right thing to do is for the topic and handle to both have their own Equal() method

**Comment by willscott** [comment]
added an `Equal` method. go convention is that this is of the form `libtalek.Equal(a, b *Handle)`, not `handle.Equal(other)`.  aliasing rules also mean we only get one of those per package, so i went with handle rather than topic.

**Comment by ryscheng** [comment]
:+1:

**Pull request merged**